### PR TITLE
[IMM32] Source code split and restructure!

### PR DESCRIPTION
**Pull request by katahiromz**
## Purpose
Implementing Japanese input...
JIRA issue: [CORE-11700](https://jira.reactos.org/browse/CORE-11700)

## Proposed changes

- Restructure due to source file bloating!

### CODE REVIEW

File: win32ss/include/ntuser.h [hunk 1]
```diff
@@ -1289,21 +1289,22 @@ C_ASSERT(sizeof(IMEDPI) == 0xa8);
 /* unconfirmed */
 typedef struct tagCLIENTIMC
 {
-    HIMC hImc;
+    HLOCAL hLocalInputContext;
```
Comment (HBelusca) [inline win32ss/include/ntuser.h:1292]:
```suggestion
    HANDLE hInputContext;
```

File: win32ss/include/ntuser.h [hunk 2]
```diff
@@ -1289,21 +1289,22 @@ C_ASSERT(sizeof(IMEDPI) == 0xa8);
...
     RTL_CRITICAL_SECTION cs;
-    DWORD unknown2;
+    UINT uCodePage;
     HKL hKL;
     BOOL bUnknown4;
 } CLIENTIMC, *PCLIENTIMC;
 
 #ifndef _WIN64
-C_ASSERT(offsetof(CLIENTIMC, hImc) == 0x0);
+C_ASSERT(offsetof(CLIENTIMC, hLocalInputContext) == 0x0);
```
Comment (HBelusca) [inline win32ss/include/ntuser.h:1303]:
```suggestion
C_ASSERT(offsetof(CLIENTIMC, hInputContext) == 0x0);
```

File: sdk/include/reactos/wine/ddk/imm.h [hunk 3]
```diff
@@ -96,13 +96,15 @@ typedef struct INPUTCONTEXTDX /* unconfirmed */
     DWORD dwUIFlags;
     DWORD dwUnknownDog;
     void *pUnknownFox;
-    /* ... */
+    DWORD dwUnknownPig;
+    DWORD dwUnknownPony;
```
Comment (HBelusca) [inline sdk/include/reactos/wine/ddk/imm.h:100]: come on.... Please use standard names `dwUnknown1`, `dwUnknown2` and so on.

File: modules/rostests/apitests/imm32/clientimc.c [hunk 4]
```diff
@@ -24,15 +24,15 @@ START_TEST(clientimc)
     DWORD dwCode;
     CLIENTIMC *pClientImc = HeapAlloc(GetProcessHeap(), HEAP_ZERO_MEMORY, sizeof(CLIENTIMC));
 
-    pClientImc->hImc = (HIMC)ImmCreateIMCC(4);
+    pClientImc->hLocalInputContext = (HLOCAL)ImmCreateIMCC(4);
```
Comment (HBelusca) [inline modules/rostests/apitests/imm32/clientimc.c:27]: Just use `hInputContext` (here and below).

File: dll/win32/imm32/imm32.c [hunk 5]
```diff
@@ -0,0 +1,2379 @@
...
+            hKL = NULL;
+    }
+
+    RegCloseKey(hLayoutKey);
+    if (hLayoutsKey)
+        RegCloseKey(hLayoutsKey);
+    return hKL;
+}
+
+typedef struct _tagImmHkl{
```
Comment (HBelusca) [inline dll/win32/imm32/imm32.c:106]:
```suggestion
typedef struct _tagImmHkl
{
```

File: dll/win32/imm32/imm32.c [hunk 6]
```diff
@@ -0,0 +1,2379 @@
...
+typedef struct tagInputContextData
+{
+        DWORD           dwLock;
+        INPUTCONTEXT    IMC;
+        DWORD           threadID;
+
+        ImmHkl          *immKbd;
+        UINT            lastVK;
+        BOOL            threadDefault;
+        DWORD           magic;
```
Comment (HBelusca) [inline dll/win32/imm32/imm32.c:143]: You may take the opportunity to fix the indentation there.

File: dll/win32/imm32/precomp.h [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+#include <strsafe.h>
+
+#define IMM_INIT_MAGIC 0x19650412
+#define IMM_INVALID_CANDFORM ULONG_MAX
+#define INVALID_HOTKEY_ID 0xFFFFFFFF
+#define MAX_CANDIDATEFORM 4
+
+#define LANGID_CHINESE_SIMPLIFIED MAKELANGID(LANG_CHINESE, SUBLANG_CHINESE_SIMPLIFIED)
+#define LANGID_CHINESE_TRADITIONAL MAKELANGID(LANG_CHINESE, SUBLANG_CHINESE_TRADITIONAL)
+#define LANGID_JAPANESE MAKELANGID(LANG_JAPANESE, SUBLANG_DEFAULT)
```
Comment (HBelusca) [inline dll/win32/imm32/precomp.h:33]:
```suggestion
#define LANGID_CHINESE_SIMPLIFIED   MAKELANGID(LANG_CHINESE, SUBLANG_CHINESE_SIMPLIFIED)
#define LANGID_CHINESE_TRADITIONAL  MAKELANGID(LANG_CHINESE, SUBLANG_CHINESE_TRADITIONAL)
#define LANGID_JAPANESE             MAKELANGID(LANG_JAPANESE, SUBLANG_DEFAULT)
```

File: dll/win32/imm32/imm32.c (hunk 6 above)
Comment (katahiromz) [reply dll/win32/imm32/imm32.c:143]: This code is useless.

File: dll/win32/imm32/imp.c [hunk 8]
```diff
@@ -0,0 +1,208 @@
...
+ *              Copyright 2018 Amine Khaldi <amine.khaldi@reactos.org>
+ *              Copyright 2020 Oleg Dubinskiy <oleg.dubinskij2013@yandex.ua>
+ *              Copyright 2020-2021 Katayama Hirofumi MZ <katayama.hirofumi.mz@gmail.com>
+ */
+
+#include "precomp.h"
+
+WINE_DEFAULT_DEBUG_CHANNEL(imm);
+
+#ifdef IMP_SUPPORT /* 3.x support */
```
[the remark this answers is not in the document]
Comment (katahiromz) [reply dll/win32/imm32/imp.c:17]: My mistake of input method profiler. This should be 'NT3'.

[the remark this answers is not in the document]
Comment (HBelusca) [reply dll/win32/imm32/imp.c:17]: > input method profiler

This full name should then be mentioned in the description in the file "header comment".

File: win32ss/include/ntuser.h (hunk 1 above)
Comment (katahiromz) [reply win32ss/include/ntuser.h]: OK.

File: win32ss/include/ntuser.h (hunk 2 above)
Comment (katahiromz) [reply win32ss/include/ntuser.h]: OK.

File: sdk/include/reactos/wine/ddk/imm.h (hunk 3 above)
Comment (katahiromz) [reply sdk/include/reactos/wine/ddk/imm.h]: OK.

File: modules/rostests/apitests/imm32/clientimc.c (hunk 4 above)
Comment (katahiromz) [reply modules/rostests/apitests/imm32/clientimc.c]: OK.

File: dll/win32/imm32/imm32.c (hunk 5 above)
Comment (katahiromz) [reply dll/win32/imm32/imm32.c]: OK.

File: dll/win32/imm32/imm32.c (hunk 6 above)
Comment (katahiromz) [reply dll/win32/imm32/imm32.c]: OK.

**Comment by HBelusca** [comment]
@katahiromz I see also some intermediate commits that don't seem to be related at all to the main point of this PR that is, just code refactoring (one should not mix that with functional code changes/unrelated fixes!)

So I think that also these three points (from the description)
> Add ValidateHandleNoErr, Imm32InitInputContext and Imm32LockIMCEx helper functions.
> Rewrite ImmLockIMC and Imm32CleanupContext functions by using the helper functions.
> Modify CLIENTIMC and INPUTCONTEXTDX structures.

should be done in a separate PR once the source code refactoring has been done.

File: dll/win32/imm32/precomp.h (hunk 7 above)
Comment (katahiromz) [reply dll/win32/imm32/precomp.h]: OK.

File: dll/win32/imm32/precomp.h [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+#include "../../../win32ss/include/ntuser.h"
+#include "../../../win32ss/include/ntwin32.h"
+#include <undocuser.h>
+#include <imm32_undoc.h>
+#include <strsafe.h>
+
+#define IMM_INIT_MAGIC 0x19650412
+#define IMM_INVALID_CANDFORM ULONG_MAX
+#define INVALID_HOTKEY_ID 0xFFFFFFFF
+#define MAX_CANDIDATEFORM 4
```
[the remark this answers is not in the document]
Comment (katahiromz) [reply dll/win32/imm32/precomp.h]: OK.

File: dll/win32/imm32/imm32.c [hunk 10]
```diff
@@ -0,0 +1,2379 @@
...
+
+static CRITICAL_SECTION threaddata_cs;
+static CRITICAL_SECTION_DEBUG critsect_debug =
+{
+    0, 0, &threaddata_cs,
+    { &critsect_debug.ProcessLocksList, &critsect_debug.ProcessLocksList },
+      0, 0, { (DWORD_PTR)(__FILE__ ": threaddata_cs") }
+};
+static CRITICAL_SECTION threaddata_cs = { &critsect_debug, -1, 0, 0, 0, 0 };
+
```
[the remark this answers is not in the document]
Comment (katahiromz) [reply dll/win32/imm32/imm32.c:167]: Deleted.

File: dll/win32/imm32/imp.c (hunk 8 above)
[the remark this answers is not in the document]
Comment (katahiromz) [reply dll/win32/imm32/imp.c]: OK.

**Comment by katahiromz** [comment]
@HBelusca Reverted some code.

**Comment by katahiromz** [comment]
Check time. I will wait for 7 days.

**Comment by katahiromz** [comment]
Merging is postponed due to https://jira.reactos.org/browse/CORE-17741 .

**Pull request closed without merging**